File: C++/source/Framework/Messaging/LightHouse/LightPulseDispatchObjectManager.cpp
/***************************************************************************
 *   Copyright (C) 20016 Vidyasagara Guntaka                               *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Messaging/LightHouse/LightPulseDispatchObjectManager.h"
#include "Framework/Messaging/LightHouse/LightPulseDispatchWorker.h"
#include "Framework/Messaging/LightHouse/LightPulseRegistrationWorker.h"
#include "Framework/Messaging/LightHouse/LightPulseUnregistrationWorker.h"

namespace WaveNs
{

LightPulseDispatchObjectManager::LightPulseDispatchObjectManager ()
    : WaveLocalObjectManager (getServiceName ())
{
    m_pLightPulseDispatchWorker = new LightPulseDispatchWorker (this);

    waveAssert (NULL != m_pLightPulseDispatchWorker, __FILE__, __LINE__);

    m_pLightPulseRegistrationWorker = new LightPulseRegistrationWorker (this);

    waveAssert (NULL != m_pLightPulseRegistrationWorker, __FILE__, __LINE__);

    m_pLightPulseUnregistrationWorker = new LightPulseUnregistrationWorker (this);

    waveAssert (NULL != m_pLightPulseUnregistrationWorker, __FILE__, __LINE__);
}

LightPulseDispatchObjectManager::~LightPulseDispatchObjectManager ()
{
    if (NULL != m_pLightPulseDispatchWorker)
    {
        delete m_pLightPulseDispatchWorker;
    }

    if (NULL != m_pLightPulseRegistrationWorker)
    {
        delete m_pLightPulseRegistrationWorker;
    }

    if (NULL != m_pLightPulseUnregistrationWorker)
    {
        delete m_pLightPulseUnregistrationWorker;
    }
}

string LightPulseDispatchObjectManager::getServiceName ()
{
    return ("Light Pulse Receiver Dispatch");
}

LightPulseDispatchObjectManager *LightPulseDispatchObjectManager::getInstance ()
{
    static LightPulseDispatchObjectManager *pLightPulseDispatchObjectManager = new LightPulseDispatchObjectManager ();

    WaveNs::waveAssert (NULL != pLightPulseDispatchObjectManager, __FILE__, __LINE__);

    return (pLightPulseDispatchObjectManager);
}

WaveServiceId LightPulseDispatchObjectManager::getWaveServiceId ()
{
    return ((getInstance ())->getServiceId ());
}

}

File: Java/source/com/CxWave/Wave/Framework/Messaging/LightHouse/LightPulseDispatchObjectManager.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Messaging.LightHouse;

import com.CxWave.Wave.Framework.ObjectModel.WaveLocalObjectManager;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.ObjectManagerPriority;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Resources.ResourceEnums.WaveObjectManagerPriority;

@ObjectManagerPriority (WaveObjectManagerPriority.WAVE_OBJECT_MANAGER_PRIORITY_LIGHT_HOUSE_DISPATCHER)
public class LightPulseDispatchObjectManager extends WaveLocalObjectManager
{
    private static LightPulseDispatchObjectManager m_lightPulseDispatchObjectManager = null;

    public LightPulseDispatchObjectManager ()
    {
        super (getServiceName ());
    }

    public static String getServiceName ()
    {
        return ("Light House Dispatch");
    }

    public static LightPulseDispatchObjectManager getInstance ()
    {
        if (null == m_lightPulseDispatchObjectManager)
        {
            m_lightPulseDispatchObjectManager = new LightPulseDispatchObjectManager ();
        }

        WaveAssertUtils.waveAssert (null != m_lightPulseDispatchObjectManager);

        return (m_lightPulseDispatchObjectManager);
    }

    public static WaveServiceId getWaveServiceId ()
    {
        return ((getInstance ()).getServiceId ());
    }
}
